### ISSUE TITLE
RFC: ipv6

### ISSUE THREAD

**Issue by pvizeli** [home-assistant/supervisor#2133]
## Context

Docker IPv6 support does not really exist and not usable because most users have an IPv6 host network from the provider without routing. Docker itself is driven by the big cloud provider which has very limited IPv6 support.

However, our system should work in IPv6 and IPv4 in the same way.

## Decision

We add a new plugin called `ipv6` which implement NAT6 -> https://github.com/robbertkl/docker-ipv6nat
Next, we need to take an IPv6 address alias over NetworkManager and attach it to the DNS plugin for support ipv6 DNS server.

**Comment by jhampson-dbre** [comment]
Since this was originally opened, the docker-ipv6nat repo mentioned above has pinned an [issue considering deprecation](https://github.com/robbertkl/docker-ipv6nat/issues/65) due to recent progress in Docker for built in IPv6 NAT support.

**Comment by pvizeli** [comment]
Nice, so we can update libnetwork on our OS

**Comment by TimoFriedri** [comment]
Hi,

the issue mentioned here is probably the reason for addons like Wireguard not to work. 
I have an IPv6 only provider. I can transmit to the Wireguard plugin from my phone (I see that on the sent and received data) BUT WG cannot answer via IPv6. So the phone never receives anything. --> no handshake

So I think this issue is highly relevant for a growing amount of people, since Dual Stack-Lite is more and more common.

**Comment by pvizeli** [comment]
It comes in 1-2 years. Docker/Moby is used and developed by some big cloud company that has self not good IPv6 support. But of the curse, the need comes and they start to fix their product which allows us to improve our software.

State today, IPv6 only installation doesn't work, you need to make dual-stack support on your network and tun it to an ipv4 gateway.

**Comment by TimoFriedri** [comment]
Uff, that's a bummer, 
Basically means an extra ~5Euro per month for a lot of people to their provider.

I think I have to run wireguard outside the docker environment directly on the machine, not nice, but managable.
Thanks for the feedback.

**Comment by tmoore22** [comment]
+1 to this, its 2023 now, does docker properly support ipv6 networking now?

**Comment by g00gleit** [comment]
+1 to this I recently switched to T-Mobile and it's annoying that I can't setup IPV6 to work with the Wireguard server on home assistant.

**Comment by bikeshedder** [comment]
Not being able to use IPv6 is a real bummer to me as well. I'm seriously considering kicking HASSOS in favor of a custom container installation or setting up my own reverse proxy.

I do love the easy updates HASSOS provides. That's the only reason that I haven't switched, yet.

The lack of IPv6 support almost looks like a planned oversight in order to push the Home Assistant Cloud service which shouldn't be a requirement to access HASS installations on a IPv6 only network.

**Comment by agners** [comment]
> Not being able to use IPv6 is a real bummer to me as well.

@bikeshedder just to be clear: Home Assistant OS supports IPv6 just fine! You can also connect to Home Assistant using IPv6 just fine! Home Assistant Core uses host network, and we do support IPv6 on the main network via Network Manager.

This issue is about supporting IPv6 for add-ons which do *not* run on host network. The Supervisor (or rather the Docker container engine) provides an isolated network for add-ons. This isolated network uses IPv4 currently. This typically is no big deal as almost all setups still use dual-stack (IPv4 and IPv6). So, if necessary, the add-ons can connect to the Internet via IPv4 still. This whole issue is really only relevant if you run a IPv6 *only* network.

**Comment by bikeshedder** [comment]
The problem I run into is the carrier grade NAT from my provider. Yes, add-ons can still access IPv4 services. It cannot use services which are only exposed via IPv6 though. This breaks the DuckDNS add-on if I want to expose my HASS Installation via IPv6 as it can't reliably detect the public facing IPv6 address.

Dedicated public facing IPv4 addresses are becoming less common in Germany as the IPv4 address pools are well known to be depleted for quite some time now.

IPv6 was introduced 27 years ago and is the only solution to end that CGN madness once and for all.

What needs to be done in order to enable IPv6 for add-ons?

**Comment by TimoFriedri** [comment]
> Hi,
> 
> the issue mentioned here is probably the reason for addons like Wireguard not to work. I have an IPv6 only provider. I can transmit to the Wireguard plugin from my phone (I see that on the sent and received data) BUT WG cannot answer via IPv6. So the phone never receives anything. --> no handshake
> 
> So I think this issue is highly relevant for a growing amount of people, since Dual Stack-Lite is more and more common.

I'd like to quote myself here. IPv6 internet providers are a common thing nowadays!

`This whole issue is really only relevant if you run a IPv6 only network.`

Many do unintentionally, because the provider only allows full IPv6 and crippled IPv4. 

MY only solution was to move Wireguard out of the Home Assistant Machine and run it seperately.

**Comment by agners** [comment]
> Many do unintentionally, because the provider only allows full IPv6 and crippled IPv4.

I guess you talk about CGNAT? That use case is handled well by the current implementation since the add-ons only reach out via IPv4, which works through CGNAT.

> My only solution was to move Wireguard out of the Home Assistant Machine and run it separately.

Yeah that add-on might benefit from full IPv6. However, I guess what you want is to have it reachable by IPv6? In that case we would need to assign the network public IPv6 addresses, which not in all IPv6 setups might be necessary.

Btw, the Wireguard add-on decides to *not* run on host network, which makes it more a Wireguard "client" (I know there is no such thing, but from an architecture more like a typical VPN client in the sense of not be reachable from the outside). That is a decision decision of that add-on.

**Comment by cleopold73** [comment]
Ran into this issue while trying to get more stuff working over ipv6 in my internal network.  In my case it was the unifi addon talking to the unifi controller.  It still work since I'm dual stack, but i was a bit disappointed that add-ons can't talk outbound to anything via ipv6.

**Comment by ShadowJonathan** [comment]
I'm going to assume that HAOS runs docker, in which case you can go looking here for ipv6 support: https://docs.docker.com/config/daemon/ipv6/

However, it's tagged as "experimental", which doesn't inspire much confidence

**Comment by Giga-Pudding** [comment]
According to the [official documentation](https://docs.docker.com/config/daemon/ipv6/), assigning private IPv6 addresses (ULA) to the docker containers is possible. IPv6 NAT is also possible.

Here's a good summary about the current status: https://github.com/robbertkl/docker-ipv6nat/issues/65#issuecomment-1568033924
Although... this comment is now 8 months old. IPv6 support may be even better today.

Does anyone see a reason, why it shouldn't be possible to use IPv6 inside HA OS docker containers?

> However, it's tagged as "experimental", which doesn't inspire much confidence

Yes, it still requires the experimental flag, but this parameter is enabled in HA OS anyway.

**Comment by deviantintegral** [comment]
I think it's (unfortunately) still important to consider and have checks for broken IPv6 connectivity with fallbacks for addons.

A case in point: Bell Canada's mobile internet for rural service has been provisioning IPv6 addresses for years. While those addresses are global addresses, they don't actually support routing traffic over IPv6. This recently came up setting up Home Assistant Cloud, where setup fails until you disable IPv6 in HAOS.

This isn't an issue with typical desktop or mobile systems because they have some sort of IPv6 connectivity test and fallback, such as with https://en.wikipedia.org/wiki/Happy_Eyeballs.

**Comment by bdraco** [comment]
aiohttp doesn't support happyeyeballs yet, but will in 3.10 it will after https://github.com/aio-libs/aiohttp/pull/7954 

Enabling ipv6 is likely not a good idea if there are any aiohttp requests until we upgrade to 3.10 since there will be unexpected breakage otherwise.

**Comment by agners** [comment]
> I think it's (unfortunately) still important to consider and have checks for broken IPv6 connectivity with fallbacks for addons.

Well but the same could be said for Home Assistant Core and Supervisor itself. :man_shrugging: 

In fact, most add-ons probably don't use aiohttp. However, if whatever they use implements happy eyeballs support is a different question of course :see_no_evil: 

I do understand the concern, but I think this mainly adds the requirement that users should be able to disable the feature if necessary. We probably could tie it to the host network configuration (e.g. if there is no host network IPv6 configured, then we'd also not provide IPv6 addresses for add-ons). Having an internet routable IPv6 address is presumably the minimum requirement for NAT66 to be useful :thinking:

**Comment by agners** [comment]
I've started a related discussion about `macvlan` support in Supervisor, see https://github.com/home-assistant/architecture/discussions/1034.

As the title of this discussion says, this issue is about IPv6 for network isolated add-ons. Introducing `macvlan` support would add another way for add-ons to get IPv6 support without using host network. I also think that `macvlan` would work very well together with IPv6: `macvlan` interfaces need IP addresses assigned. This is much simpler in IPv6 world thanks to SLAAC.

**Comment by mraerino** [comment]
quick note that i got this working on an out-of-the-box Homeassistant OS install this way. Maybe it can be an inspiration for how to solve it natively:

<details>
  <summary>Enabling IPv6 for addon containers</summary>

You need to run all of these commands on a native SSH session. The SSH addon won't work.

1. Stop supervisor: `systemctl stop hassos-supervisor`
2. Stop all running containers:
    ```
    for ID in $(docker ps --format '{{ .ID }}'); do docker stop $ID; done
    ```
3. Remove all containers:
    ```
    for ID in $(docker ps -a --format '{{ .ID }}'); do docker rm $ID; done
    ```
4. Remove the `hassio` network
    ```
    docker network rm hassio
    ```
5. Re-create the `hassio` network with IPv6 support enabled
    ```
    docker network create hassio -d bridge --ipv6 --subnet <YOUR ULA NET> --subnet 172.30.32.0/23 --ip-range 172.30.33.0/24 --gateway 172.30.32.1 --opt com.docker.network.bridge.name=hassio
    ```
    **Remember to insert your ULA net!**
6. Reboot!

</details>

This had the consequence of giving all addon containers an IPv6 address that works for outgoing connections.

The only trouble i had was that the `core_nginx_proxy` addon does not listen on IPv6 by default, which i was able to work around but would need to be fixed before IPv6 is rolled out to everyone.

The nice consequence is that now trusted networks work with IPv6 client when using the nginx addon.

**Comment by agners** [comment]
@mraerino I guess this is more or less what #3780 implements then, no?

What I am a bit worried is how add-ons typically behave in environments which do not have (global) IPv6 connectivity. From what I understand this adds NAT66. And the approach makes software inside the add-ons think IPv6 is available, in any case. Techniques such as Happy eyeballs should make sure to fallback to IPv4 quickly, but this might lead to slowdowns or issues with happy eyeballs is not beeing used.

**Comment by Giga-Pudding** [comment]
Can't we let the users decide on their own? Let's have a switch in the network config:
 "enable upstream IPv6 support (IPv6 support by ISP mandatory)"

This should be disabled by default, so everything is fine.

**Comment by codyc1515** [comment]
I would strongly encourage enabling by default. However, is there a security risk from this? Many people using IPv4 rely on NAT as a form of firewall - IPv6 does not (usually) have this.

**Comment by Giga-Pudding** [comment]
In this case, private IPv6 addresses are used (unique local addresses). Docker performs a NAT66, so what happens is basically the same as with IPv4.

Exposed hosts do not exist by default. The containers shouldn't be reachable from outside, without proper iptables rules.

**Comment by mraerino** [comment]
@agners i think the proposal we've landed on in this issue seems reasonable. what is the process for making it happen?

**Comment by Giga-Pudding** [comment]
@mraerino can you do me a favor and test something?
What happens, if IPv6 is enabled for containers like you did it, but the supervised host does not have a working IPv6 upstream for some reason? How does Docker behave in that case? Will Containers still think there's a working IPv6 upstream and try to send traffic via IPv6 into the internet? Or is Docker clever enough to remove the IPv6 default gateway for the containers?

Can you set the IPv6 configuration in your running Home Assistant OS to "disabled" and see what happens?

**Comment by agners** [comment]
I have the same concern as @Giga-Pudding has here: How do the add-ons behave in absence of IPv6 support on the host side...

@mraerino I don't quite understand your setup at this point. In your case, what is "your ULA net"? Is that the same ULA you use on your regular LAN? So you essentially you bridge the IPv6 network so the add-on are directly in your LAN? If so then there is no NAT66 in play on the Home Assistant side? But how can you then reach the public internet, is NAT66 active on your router? :thinking:

**Comment by mraerino** [comment]
> In your case, what is "your ULA net"? Is that the same ULA you use on your regular LAN? So you essentially you bridge the IPv6 network so the add-on are directly in your LAN?

no, i just created a fresh ULA prefix (using https://unique-local-ipv6.com/) and used it. NAT66 in Docker is used. if i had prefix delegation enabled in my router i might have used that to get public addresses, but for now i use NAT.

**Comment by Giga-Pudding** [comment]
@mraerino can you test this?

> @mraerino can you do me a favor and test something? What happens, if IPv6 is enabled for containers like you did it, but the supervised host does not have a working IPv6 upstream for some reason? How does Docker behave in that case? Will Containers still think there's a working IPv6 upstream and try to send traffic via IPv6 into the internet? Or is Docker clever enough to remove the IPv6 default gateway for the containers?
> 
> Can you set the IPv6 configuration in your running Home Assistant OS to "disabled" and see what happens?

**Comment by Giga-Pudding** [comment]
@agners there's no progress here. Can we at least introduce full IPv6 support via NAT66 as experimental feature (opt-in)?

**Comment by benniju** [comment]
Is there any progress on this? I would really like IPv6 support for different reasons for example for monitoring IPv6 addresses. (https://github.com/hassio-addons/repository/issues/618) I think it's pretty much a no go to not properly support IPv6 in 2024.

**Comment by Meister1977** [comment]
+1
I need ipv6 support for multiple reason.
Would be nice if HA enable this feature... Please...
Maybe, you can create 2 docker networks? Keep the hassio and and a hassiov4v6 and I as a user can assign the new network? It would be backward compatible?

**Comment by ShadowJonathan** [comment]
IPv6 Private/Dynamic Addresses + Firewall is functionally equivalent to IPv4 + NAT

**Comment by lorinc-ambrozy-work** [comment]
+1 Please, we need this feature.

**Comment by Eisbergsalat** [comment]
I recently got infected by DS-lite....
And now I even have to change my Uptime Kuma from HA to docker? 😢 

I hope everyone else can stand tall during the DS-lite pandemic.
In hope you wont get infected 🫡

**Comment by gtxaspec** [comment]
+1 Its 2024, IPv6 is needed by default.

**Comment by PauloHeaven** [comment]
OK, so that's why. I've got an IPv6-only infrastructure because IPv4 addresses become more and more expensive, and missing IPv6 support prevents the Let's Encrypt module from reaching my authoritative DNS server to acquire certificates.

**Comment by Eisbergsalat** [comment]
> I recently got infected by DS-lite.... And now I even have to change my Uptime Kuma from HA to docker? 😢
> 
> I hope everyone else can stand tall during the DS-lite pandemic. In hope you wont get infected 🫡


FunFact:
It took about 10 days to convert everything in my HomeLab to full IPv6-compatibility. Switched from DuckDNS to IPv64, Disabled IPv4 and everything seemed to work! (IPv6 Only! No Ipv4!)

a few days later my wife complained, that she couldnt access Home Assistant, while using her mobile Internet over SIM...
And Lo and Behold: Aldi Talk doesnt support IPv6 at all ... 
I wrote them about 2 months ago. No answer.
Luckily Vodafone was gracious enough to lift my "lite" Curse, so I am back to DS.

Jokes aside: IPv6 in 2024 almost 2025 is a must.

**Comment by skgsergio** [comment]
I've been fighting a couple of days with IPv6 for addons (especially for Uptime Kuma) just to see the PR linked above was closed 2 days ago, sad: https://github.com/home-assistant/supervisor/pull/3780#event-16389841411

**Comment by davidrapan** [comment]
> PR linked above was closed 2 days ago, sad: https://github.com/home-assistant/supervisor/pull/3780#event-16389841411

It has been implemented in https://github.com/home-assistant/supervisor/pull/5879

And is configurable using `ha docker options --enable-ipv6=true`

**Comment by benniju** [comment]
Thats awesome, at first I thought that it didn't work, but I just had to reboot the Full HA System and not just the containers (makes sense in hindsight).
Now it works well, for example with a ping to ipv6.google.com in Uptime Kuma.

**Comment by davidrapan** [comment]
Yes, I forgot to mention that a host restart is required, but that's the same as the manual method (and the log will inform you).

**Comment by skgsergio** [comment]
@davidrapan if we did the manual steps... we can run the command? we should rollback the changes and run the command? or just keep as we are?

Edit: I ran the command without touching my setup and restarted and now info says that ipv6 is enabled but using the ULA I set not the one used in the PR but it works. I might try recreating the network at some point to remain compatible with the official setup.

**Comment by davidrapan** [comment]
And I would like to add that the IPv6 plugin mentioned by the op is no longer needed, I think (but correct me if I'm wrong), as all the issues have been resolved in the meantime.

**Comment by agners** [comment]
Yeah I think this can basically be closed. I'll double check tomorrow if we are complete on this one.

> ... (and the log will inform you).

So about that: It seems *not* to inform through logs when you just set the option. We only log when actually trying to migrate, which is on Supervisor (re)start. 

Ideally we'd show something in the CLI. I don't think we have a "success with message" logic in our API. For now, I'd be ok by just hacking it into the CLI (when docker options with `--enable-ipv6` is called print a hint that a system restart is required for it to take effect).

**Comment by skgsergio** [comment]
> And I would like to add that the IPv6 plugin mentioned by the op is no longer needed

Is not needed, I never used it in fact.

**Comment by davidrapan** [comment]
> So about that

Yeah, I meant between the lines, like... "will let you know if the restart you did wasn't enough..."

> is called print a hint that a system restart is required for it to take effect

That would be a neat addition indeed.

**Comment by agners** [comment]
With #5879 and the default for new installations enabling IPv6 support with #6029, this is resolved.

We probably want to enable IPv6 support by default for existing users at one point, but I rather like to have some more experience in this domain.

In my testing, IPv6 connectivity works just fine if explicitly asked for, e.g. `curl` or `ping` default to IPv4 within a add-on, presumably because there is only a ULA network address available:

```
# ping -6 www.google.ch
PING www.google.ch(zrh04s15-in-x03.1e100.net (2a00:1450:400a:803::2003)) 56 data bytes
64 bytes from zrh04s15-in-x03.1e100.net (2a00:1450:400a:803::2003): icmp_seq=1 ttl=108 time=2.22 ms
```

On networks which do not have IPv6 available, and if explicitly IPv6 connection is requested, the expected network level errors are raised:
```
# ping -6  www.google.ch
PING www.google.ch(zrh04s15-in-x03.1e100.net (2a00:1450:400a:803::2003)) 56 data bytes
From fd0c:ac1e:2100::1 icmp_seq=1 Destination unreachable: No route
```

```
# curl -v -6 www.google.ch
*   Trying [2a00:1450:400a:803::2003]:80...
* connect to 2a00:1450:400a:803::2003 port 80 failed: Network is unreachable
```

I guess for "true"/full IPv6 connectivity, Docker would need support for IPv6 prefix delegation, and it only would work for users who actually have prefix delegation available and enabled. But I consider this support out of scope for this issue.

**Comment by lolo32** [comment]
Hi,

I just found this thread, and just before trying the solution I found a solution that worked and enabled IPv6 for Uptime-Kuma and the others addons too.

Yes, it works ✅

1. I installed the`Advanced SSH & Web Terminal` module,
2. opened it,
3. I launched the command `ha docker options --enable-ipv6=true`
4. after that, Home Assistant said that a configuration change need to restart
5. and that's all, everything work like a charm

I just do not understand why this command is not launched automatically these days
ℹ️ Note: this is not to expose addons using IPv6 addresses, this option enable IPv6 routing using private addresses

**Comment by agners** [comment]
> I just do not understand why this command is not launched automatically these days

The options just got added a couple weeks back with #5879. PR #6029 makes it the default for new installations, however it is only enabled by default on the very first boot. The current Home Assistant release 16.0 comes with a Supervisor before that release, so that is why it is still off by default.

However, with the upcoming HAOS 16.1 release IPv6 will be enabled by default for new installations.

Once that proves to cause no negative effect in wider deployment, we could also enable it by default for all installations. But I am a bit hesitant on flipping that switch as it will affect essentially all users at once 🥶

**Comment by benniju** [comment]
I agree that it should probably not be enabled by default afterwards for now. But I think it would be good if the setting cloud be changed in the UI.

**Comment by lolo32** [comment]
> The options just got added a couple weeks back with [#5879](https://github.com/home-assistant/supervisor/pull/5879). PR [#6029](https://github.com/home-assistant/supervisor/pull/6029) makes it the default for new installations, however it is only enabled by default on the very first boot. The current Home Assistant release 16.0 comes with a Supervisor before that release, so that is why it is still off by default.

I do not understand, I installed a new fresh installation of HomeAssistant the July, 29th, but this option was disabled, it was not enabled like you said @agners

**Comment by benniju** [comment]
> I do not understand, I installed a new fresh installation of HomeAssistant the July, 29th, but this option was disabled, it was not enabled like you said [@agners](https://github.com/agners)

As @agners wrote in the post you quoted, it is still not enabled in the current release of Home Assistant.
